<a href="https://colab.research.google.com/github/innovinitylabs/learning-pytorch/blob/main/00_pytorch_fundamentals.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [84]:
import torch

In [85]:
# mps_device = torch.device('mps')

In [86]:
# class Conf: dev = torch.device("mps")

# introduction to tensors

## creating tensors

pytorch tensors are created using `torch.tensors`

### Scalar

In [87]:
scalar = torch.tensor(7)
scalar

tensor(7)

to get dimensions

In [88]:
scalar.ndim

0

to get the number without type

In [89]:
scalar.item()

7

### Vector

In [90]:
vector = torch.tensor([7, 7])
vector

tensor([7, 7])

In [91]:
vector.ndim

1

easy way to remember dimension is by number of square brackets

In [92]:
vector.shape

torch.Size([2])

shape gives how elements in tensor are arranged.

### MATRIX

In [93]:
MATRIX = torch.tensor([[7, 8],
                       [9,10]])
MATRIX

tensor([[ 7,  8],
        [ 9, 10]])

In [94]:
MATRIX.ndim

2

In [95]:
MATRIX.shape

torch.Size([2, 2])

In [96]:
MATRIX[0]

tensor([7, 8])

### TENSOR

In [97]:
TENSOR = torch.tensor([[[1, 2, 3],
                        [3, 6, 9],
                        [2, 4, 5]]])
TENSOR

tensor([[[1, 2, 3],
         [3, 6, 9],
         [2, 4, 5]]])

In [98]:
TENSOR.ndim

3

In [99]:
TENSOR.shape

torch.Size([1, 3, 3])

In [100]:
TENSOR[0]


tensor([[1, 2, 3],
        [3, 6, 9],
        [2, 4, 5]])

### Random Tensors

random tensors are important because many Neural networks starts with random numbers

In [101]:
random_tensor = torch.rand(3, 4)
random_tensor

tensor([[0.4521, 0.7020, 0.6274, 0.0945],
        [0.8525, 0.3572, 0.7492, 0.3579],
        [0.5453, 0.8171, 0.0570, 0.4560]])

In [102]:
random_tensor.ndim

2

In [103]:
random_huge_tensor = torch.zeros(4,5)
random_huge_tensor

tensor([[0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.]])

In [104]:
random_huge_tensor.dtype

torch.float32

In [105]:
torch.__version__

'2.0.1+cu118'

In [106]:
zttendep = torch.arange(0, 10)
zttendep

tensor([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [107]:
ten = torch.tensor([11, 22, 33])

In [108]:
ten * 10

tensor([110, 220, 330])

In [109]:
torch.mul(ten, 11)

tensor([121, 242, 363])

In [110]:
ten * ten

tensor([ 121,  484, 1089])

In [111]:
twotwo = torch.tensor([[1, 2],
                       [3, 4]])

In [112]:
twotwo * twotwo

tensor([[ 1,  4],
        [ 9, 16]])

In [113]:
torch.matmul(twotwo, twotwo)

tensor([[ 7, 10],
        [15, 22]])

In [114]:
torch.matmul(ten, ten)

tensor(1694)

In [115]:
tensor1 = torch.randn(10, 3, 4)
tensor2 = torch.randn(4)
torch.matmul(tensor1, tensor2).size()

torch.Size([10, 3])

In [116]:
tensor1.size()

torch.Size([10, 3, 4])

In [117]:
tensor2.size()

torch.Size([4])

In [118]:
tensor1, tensor2

(tensor([[[-0.1915,  0.1284, -1.4183, -0.1282],
          [-0.4212, -0.9606, -0.9892,  0.0638],
          [-0.0178, -1.3208, -0.3971,  0.1244]],
 
         [[ 1.1560, -1.9490,  0.9272,  1.3656],
          [ 1.2406,  0.7299, -0.8683,  0.6726],
          [ 0.1142,  1.0251,  1.3822,  2.1080]],
 
         [[ 1.4027,  1.9010, -1.1275,  0.7207],
          [-0.2668, -1.6520, -1.1392,  0.6818],
          [-0.5700,  0.9400,  1.5548, -0.3399]],
 
         [[ 0.6953, -2.7137,  1.6436,  2.9385],
          [ 0.2241,  1.8586, -1.3166,  1.6165],
          [-0.5542, -0.6670,  0.2274,  0.8998]],
 
         [[-1.8420,  0.8163, -0.4833,  0.0640],
          [-0.3884, -0.4140, -0.7175,  0.2219],
          [-0.6163, -1.8684, -0.0034, -1.0512]],
 
         [[-0.9096,  0.4972,  0.8052, -0.0940],
          [ 0.1214,  0.4600,  0.1848,  0.4180],
          [ 0.6184, -0.4224, -0.9932, -2.0695]],
 
         [[-0.2617, -1.3285,  0.6058,  1.5191],
          [ 0.4225,  0.2282,  0.1016,  1.5571],
          [-0.8145, -1

In [119]:
torch.matmul(tensor1, tensor2)

tensor([[-1.4773e+00, -2.3476e+00, -2.6691e+00],
        [-1.4264e+00,  2.7726e-03,  7.1582e+00],
        [ 1.7362e+00, -2.8905e+00,  3.6890e+00],
        [ 1.3714e+00,  4.4260e+00,  1.3049e+00],
        [ 3.2326e+00, -7.7203e-01, -4.5695e+00],
        [ 2.8094e+00,  1.6658e+00, -6.3861e+00],
        [ 1.2933e+00,  2.7980e+00, -1.9582e+00],
        [ 3.9273e-01, -1.1843e+00,  5.5964e-02],
        [-2.1835e+00,  6.3118e-01,  5.1935e+00],
        [ 1.0057e-01,  3.2916e+00,  3.2282e-01]])

In [120]:
torch.matmul(torch.rand(3,2),torch.rand(2,3))

tensor([[0.2202, 0.3332, 0.3728],
        [0.4866, 0.7188, 0.7984],
        [0.6470, 0.7602, 0.7764]])

In [121]:
x = torch.arange(0, 100, 10)
x

tensor([ 0, 10, 20, 30, 40, 50, 60, 70, 80, 90])

In [122]:
x.argmax()

tensor(9)

In [123]:
x.reshape(5,2)

tensor([[ 0, 10],
        [20, 30],
        [40, 50],
        [60, 70],
        [80, 90]])

In [124]:
x

tensor([ 0, 10, 20, 30, 40, 50, 60, 70, 80, 90])

In [125]:
torch.reshape(x, (5,2))

tensor([[ 0, 10],
        [20, 30],
        [40, 50],
        [60, 70],
        [80, 90]])

In [126]:
x

tensor([ 0, 10, 20, 30, 40, 50, 60, 70, 80, 90])

In [127]:
torch.stack([x,x,x], dim= 1)

tensor([[ 0,  0,  0],
        [10, 10, 10],
        [20, 20, 20],
        [30, 30, 30],
        [40, 40, 40],
        [50, 50, 50],
        [60, 60, 60],
        [70, 70, 70],
        [80, 80, 80],
        [90, 90, 90]])

In [128]:
x = torch.zeros(2, 1, 2, 1, 2)
x

tensor([[[[[0., 0.]],

          [[0., 0.]]]],



        [[[[0., 0.]],

          [[0., 0.]]]]])

In [129]:
y = torch.squeeze(x)
y

tensor([[[0., 0.],
         [0., 0.]],

        [[0., 0.],
         [0., 0.]]])

In [130]:
x[:, 0] = 1
x , y

(tensor([[[[[1., 1.]],
 
           [[1., 1.]]]],
 
 
 
         [[[[1., 1.]],
 
           [[1., 1.]]]]]),
 tensor([[[1., 1.],
          [1., 1.]],
 
         [[1., 1.],
          [1., 1.]]]))

In [131]:
z = torch.unsqueeze(y, dim=1)
z

tensor([[[[1., 1.],
          [1., 1.]]],


        [[[1., 1.],
          [1., 1.]]]])

In [132]:
torch.unsqueeze(y, dim=0)

tensor([[[[1., 1.],
          [1., 1.]],

         [[1., 1.],
          [1., 1.]]]])

In [133]:
torch.unsqueeze(y, dim=1)

tensor([[[[1., 1.],
          [1., 1.]]],


        [[[1., 1.],
          [1., 1.]]]])

In [134]:
torch.unsqueeze(y, dim=3)

tensor([[[[1.],
          [1.]],

         [[1.],
          [1.]]],


        [[[1.],
          [1.]],

         [[1.],
          [1.]]]])

#### to check if squeeze shares memory

In [135]:
sqz = torch.arange(0,12, 2)
sqz

tensor([ 0,  2,  4,  6,  8, 10])

In [136]:
sqz = sqz.unsqueeze(dim=0)

In [137]:
sqz

tensor([[ 0,  2,  4,  6,  8, 10]])

In [138]:
sqz2 = sqz.squeeze( dim = 0)
sqz2

tensor([ 0,  2,  4,  6,  8, 10])

In [139]:
sqz2[1] = 69
sqz2

tensor([ 0, 69,  4,  6,  8, 10])

In [140]:
sqz

tensor([[ 0, 69,  4,  6,  8, 10]])

Reproducablity. random seed

In [141]:
r_seed = 42
torch.manual_seed(r_seed)
rand_a = torch.rand(3,2)
rand_a

tensor([[0.8823, 0.9150],
        [0.3829, 0.9593],
        [0.3904, 0.6009]])

In [142]:
torch.rand(3,2)

tensor([[0.2566, 0.7936],
        [0.9408, 0.1332],
        [0.9346, 0.5936]])

In [143]:
torch.manual_seed(42)
torch.rand(3,2)

tensor([[0.8823, 0.9150],
        [0.3829, 0.9593],
        [0.3904, 0.6009]])

In [144]:
torch.manual_seed(42)
rand_a == torch.rand(3,2)

tensor([[True, True],
        [True, True],
        [True, True]])

In [145]:
rand_a.device

device(type='cpu')

In [146]:
# mps_rans = rand_a.to(mps_device)

rand_a still stays on CPU. have to assign it to a variable to hold the migrated tensor on `mps` like above

In [147]:
# mps_rans.device

In [148]:
!nvidia-smi

/bin/bash: line 1: nvidia-smi: command not found


# Practice

using template given in course for ease

In [149]:
import torch
seed = 42

In [150]:
print(f'without setting any prior manual seed \n{torch.rand(3,3)}') #RANDOM

without setting any prior manual seed 
tensor([[0.2566, 0.7936, 0.9408],
        [0.1332, 0.9346, 0.5936],
        [0.8694, 0.5677, 0.7411]])


### torch.`random`.manual_seed()

In [151]:
torch.random.manual_seed(seed)
print(f'with random manual seed \n{torch.rand(3,3)}\n') #
print(f'call 1 \n{torch.rand(3,3)}')
print(f'call 2 \n{torch.rand(3,3)}')

with random manual seed 
tensor([[0.8823, 0.9150, 0.3829],
        [0.9593, 0.3904, 0.6009],
        [0.2566, 0.7936, 0.9408]])

call 1 
tensor([[0.1332, 0.9346, 0.5936],
        [0.8694, 0.5677, 0.7411],
        [0.4294, 0.8854, 0.5739]])
call 2 
tensor([[0.2666, 0.6274, 0.2696],
        [0.4414, 0.2969, 0.8317],
        [0.1053, 0.2695, 0.3588]])


### torch.manual_seed()

In [152]:
# torch.manual_seed(seed)
print(f'call 3 without setting seed above, call 1 if set \n{torch.rand(3,3)} \n') #
print(f'call 4 without setting seed above, call 2 if set \n{torch.rand(3,3)}') #
print(f'call 4 without setting seed above, call 3 if set \n{torch.rand(3,3)}')

with global manual seed 
tensor([[0.1994, 0.5472, 0.0062],
        [0.9516, 0.0753, 0.8860],
        [0.5832, 0.3376, 0.8090]]) 

call 1 
tensor([[0.5779, 0.9040, 0.5547],
        [0.3423, 0.6343, 0.3644],
        [0.7104, 0.9464, 0.7890]])
call 2 
tensor([[0.2814, 0.7886, 0.5895],
        [0.7539, 0.1952, 0.0050],
        [0.3068, 0.1165, 0.9103]])


In [153]:
print(f'call n \n{torch.rand(3,3)}') #totally random because it changes every time its called

call n 
tensor([[0.6440, 0.7071, 0.6581],
        [0.4913, 0.8913, 0.1447],
        [0.5315, 0.1587, 0.6542]])


In [154]:
torch.manual_seed(69)
torch.rand(3,3)

tensor([[0.8398, 0.8042, 0.1213],
        [0.5309, 0.6646, 0.4077],
        [0.0888, 0.2429, 0.7053]])

In [155]:
torch.rand(3,3)

tensor([[0.6216, 0.9188, 0.0185],
        [0.8741, 0.0560, 0.9659],
        [0.0073, 0.3628, 0.4197]])

In [156]:
torch.manual_seed(69)
torch.rand(3,3)

tensor([[0.8398, 0.8042, 0.1213],
        [0.5309, 0.6646, 0.4077],
        [0.0888, 0.2429, 0.7053]])

In [157]:
torch.rand(3,3)

tensor([[0.6216, 0.9188, 0.0185],
        [0.8741, 0.0560, 0.9659],
        [0.0073, 0.3628, 0.4197]])

In [158]:
torch.manual_seed(69)
torch.rand(3,3)
torch.rand(3,3)
torch.rand(3,3)



tensor([[0.6444, 0.0099, 0.5925],
        [0.9631, 0.6958, 0.9157],
        [0.5523, 0.2344, 0.3262]])

##### to reset the manual seed test

In [179]:
initialseed = torch.random.get_rng_state()

In [180]:
torch.manual_seed(69)
print(torch.rand(3,3))



tensor([[0.8398, 0.8042, 0.1213],
        [0.5309, 0.6646, 0.4077],
        [0.0888, 0.2429, 0.7053]])


In [181]:
print(torch.rand(3,3))

tensor([[0.6216, 0.9188, 0.0185],
        [0.8741, 0.0560, 0.9659],
        [0.0073, 0.3628, 0.4197]])


In [182]:
# print(torch.rand(3,3))

torch.random.set

In [183]:
torch.random.set_rng_state(initialseed)

In [184]:
print(torch.rand(3,3))

tensor([[0.6444, 0.0099, 0.5925],
        [0.9631, 0.6958, 0.9157],
        [0.5523, 0.2344, 0.3262]])



# print(torch.rand(3,3)) 3rd call should be
account_circle
tensor([[0.6444, 0.0099, 0.5925],
        [0.9631, 0.6958, 0.9157],
        [0.5523, 0.2344, 0.3262]])

# To reset manual seed

has to get state of the seed and save it to variable and set back.

In [191]:
initialseed = torch.random.get_rng_state()

saving initial state

In [192]:
torch.manual_seed(69)
print(f'call 1 \n{torch.rand(1,3)}')
print(f'call 2 \n{torch.rand(1,3)}')
print(f'call 3 \n{torch.rand(1,3)}')

call 1 
tensor([[0.8398, 0.8042, 0.1213]])
call 2 
tensor([[0.5309, 0.6646, 0.4077]])
call 3 
tensor([[0.0888, 0.2429, 0.7053]])


In [193]:
torch.manual_seed(69)
print(f'call 1 \n{torch.rand(1,3)}')
print(f'call 2 \n{torch.rand(1,3)}')

call 1 
tensor([[0.8398, 0.8042, 0.1213]])
call 2 
tensor([[0.5309, 0.6646, 0.4077]])


In [194]:
torch.random.set_rng_state(initialseed)

reset seed state

In [195]:
print(f'call 3 \n{torch.rand(1,3)}')

call 3 
tensor([[0.3265, 0.9798, 0.7475]])
